src/libwinpty/winpty.cc

```diff
@@ -494,24 +495,92 @@ static OwnedHandle startAgentProcess(
         sui.dwFlags |= STARTF_USESHOWWINDOW;
         sui.wShowWindow = SW_HIDE;
     }
+
     PROCESS_INFORMATION pi = {};
-    const BOOL success =
-        CreateProcessW(exePath.c_str(),
-                       cmdlineV.data(),
-                       nullptr, nullptr,
-                       /*bInheritHandles=*/FALSE,
-                       /*dwCreationFlags=*/creationFlags,
-                       nullptr, nullptr,
-                       &sui, &pi);
-    if (!success) {
-        const DWORD lastError = GetLastError();
-        const auto errStr =
-            (WStringBuilder(256)
-                << L"winpty-agent CreateProcess failed: cmdline='" << cmdline
-                << L"' err=0x" << whexOfInt(lastError)).str_moved();
-        throw LibWinptyException(
-            WINPTY_ERROR_AGENT_CREATION_FAILED, errStr.c_str());
+    if (cfg->flags & WINPTY_FLAG_IMPERSONATE_THREAD) {
+        HRESULT hr;
+        HANDLE token = nullptr;
+        if (!OpenThreadToken(
+            GetCurrentThread(),
+            TOKEN_QUERY | TOKEN_DUPLICATE | TOKEN_ASSIGN_PRIMARY,
+            FALSE,
+            &token))
+        {
+            const DWORD lastError = GetLastError();
+            const auto errStr =
+                (WStringBuilder(256)
+                    << L"winpty-agent OpenThreadToken failed: cmdline='" << cmdline
+                    << L"' err=0x" << whexOfInt(lastError)).str_moved();
+            throw LibWinptyException(
+                WINPTY_ERROR_AGENT_CREATION_FAILED, errStr.c_str());
+        }
+
+        HANDLE dupToken = nullptr;
+        if (!DuplicateTokenEx(
+            token,
+            TOKEN_QUERY | TOKEN_DUPLICATE | TOKEN_ASSIGN_PRIMARY,
+            NULL,
+            SecurityImpersonation,
+            TokenPrimary,
+            &dupToken))
+        {
+            CloseHandle(token);
+            const DWORD lastError = GetLastError();
+            const auto errStr =
+                (WStringBuilder(256)
+                    << L"winpty-agent DuplicateTokenEx failed: cmdline='" << cmdline
+                    << L"' err=0x" << whexOfInt(lastError)).str_moved();
+            throw LibWinptyException(
+                WINPTY_ERROR_AGENT_CREATION_FAILED, errStr.c_str());
+        }
+
+        DWORD exitCode = 0;
+        const BOOL success = CreateProcessAsUser(
+            dupToken,
+            exePath.c_str(),
+            cmdlineV.data(),
+            nullptr,
+            nullptr,
+            FALSE,
+            CREATE_DEFAULT_ERROR_MODE | CREATE_BREAKAWAY_FROM_JOB | CREATE_UNICODE_ENVIRONMENT | NORMAL_PRIORITY_CLASS,
+            nullptr,
+            nullptr,
+            &sui,
+            &pi);
+        if (!success) {
+            CloseHandle(token);
```

Comment:
I'd prefer this code use RAII (i.e. OwnedHandle), but I can fix that up myself after merging the change.

Reply:
left as is.